### PR TITLE
fix: exit handler variables don't get resolved correctly. Fixes #10393

### LINKED ISSUE #10393: `workflow.status` doesn't work with ternary operator in arguments
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

A following ternary operator in the exit handler`'{{= workflow.status == "Succeeded" ? "SUCCESSFUL" : "FAILED" }}'` results in "FAILED" even when workflow.status is "Succeeded".
Other workflow parameters seem to be processed correctly: I tested it with `workflow.parameters.foo` instead of status and it behaved as expected.

### Version

v3.4.4

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
spec:
  entrypoint: run-test
  templates:
    - name: run-test
      container:
        name: runner
        image: alpine
        command: ['sh','-c']
        args:
          - exit 0
    - name: cowsay
      inputs:
        parameters:
          - name: ternary
          - name: status
      container:
        image: argoproj/argosay:v1
        command: ['sh','-c']
        args:
          - cowsay "{{inputs.parameters.ternary}} {{inputs.parameters.status}}"
  hooks:
    exit:
      template: cowsay
      arguments:
        parameters:
          - name: ternary
            value: '{{= workflow.status == "Succeeded" ? "SUCCESSFUL" : "FAILED" }}'
          - name: status
            value: '{{= workflow.status }}'
```
Output logs:
```
lovely-rhino: time="2023-01-24T07:29:32.402Z" level=info msg="sub-process exited" argo=true error="<nil>"
lovely-rhino-cowsay-1502533312:  __________________
lovely-rhino-cowsay-1502533312: < FAILED Succeeded >
lovely-rhino-cowsay-1502533312:  ------------------
lovely-rhino-cowsay-1502533312:         \   ^__^
lovely-rhino-cowsay-1502533312:          \  (oo)\_______
lovely-rhino-cowsay-1502533312:             (__)\       )\/\
lovely-rhino-cowsay-1502533312:                 ||----w |
lovely-rhino-cowsay-1502533312:                 ||     ||
lovely-rhino-cowsay-1502533312: time="2023-01-24T07:29:42.139Z" level=info msg="sub-process exited" argo=true error="<nil>"
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep lovely-rhino

time="2023-01-24T07:29:28.829Z" level=info msg="Processing workflow" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:28.851Z" level=info msg="Updated phase  -> Running" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:28.851Z" level=info msg="Pod node lovely-rhino initialized Pending" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:28.897Z" level=info msg="Created pod: lovely-rhino (lovely-rhino)" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:28.897Z" level=info msg="TaskSet Reconciliation" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:28.897Z" level=info msg=reconcileAgentPod namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:28.910Z" level=info msg="Workflow update successful" namespace=test phase=Running resourceVersion=7190628 workflow=lovely-rhino
time="2023-01-24T07:29:38.899Z" level=info msg="Processing workflow" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:38.899Z" level=info msg="Task-result reconciliation" namespace=test numObjs=0 workflow=lovely-rhino
time="2023-01-24T07:29:38.899Z" level=info msg="node changed" namespace=test new.message= new.phase=Succeeded new.progress=0/1 nodeID=lovely-rhino old.message= old.phase=Pending old.progress=0/1 workflow=lovely-rhino
time="2023-01-24T07:29:38.900Z" level=info msg="TaskSet Reconciliation" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:38.900Z" level=info msg=reconcileAgentPod namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:38.900Z" level=info msg="Running OnExit handler: " namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:38.900Z" level=info msg="Pod node lovely-rhino-1502533312 initialized Pending" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:38.927Z" level=info msg="Created pod: lovely-rhino.onExit (lovely-rhino-cowsay-1502533312)" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:38.938Z" level=info msg="Workflow update successful" namespace=test phase=Running resourceVersion=7190690 workflow=lovely-rhino
time="2023-01-24T07:29:48.928Z" level=info msg="Processing workflow" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="Task-result reconciliation" namespace=test numObjs=0 workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="node changed" namespace=test new.message= new.phase=Succeeded new.progress=0/1 nodeID=lovely-rhino-1502533312 old.message= old.phase=Pending old.progress=0/1 workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="node unchanged" namespace=test nodeID=lovely-rhino workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="TaskSet Reconciliation" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg=reconcileAgentPod namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="Running OnExit handler: " namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="Updated phase Running -> Succeeded" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="Marking workflow completed" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="Checking daemoned children of " namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:48.935Z" level=info msg="cleaning up pod" action=deletePod key=test/lovely-rhino-1340600742-agent/deletePod
time="2023-01-24T07:29:48.957Z" level=info msg="Workflow update successful" namespace=test phase=Succeeded resourceVersion=7190754 workflow=lovely-rhino
time="2023-01-24T07:29:48.958Z" level=info msg="Queueing Succeeded workflow test/lovely-rhino for delete in 10m0s due to TTL"
time="2023-01-24T07:29:53.963Z" level=info msg="cleaning up pod" action=deletePod key=test/lovely-rhino/deletePod
time="2023-01-24T07:29:53.963Z" level=info msg="cleaning up pod" action=deletePod key=test/lovely-rhino-cowsay-1502533312/deletePod
```

### Logs from in your workflow's wait container

```text
lovely-rhino-f9r5l: time="2023-01-24T07:39:44.473Z" level=info msg="Starting Workflow Executor" version=v3.4.4
lovely-rhino-f9r5l: time="2023-01-24T07:39:44.477Z" level=info msg="Using executor retry strategy" Duration=1s Factor=1.6 Jitter=0.5 Steps=5
lovely-rhino-f9r5l: time="2023-01-24T07:39:44.477Z" level=info msg="Executor initialized" deadline="2023-01-24 09:39:42 +0000 UTC" includeScriptOutput=false namespace=test podName=lovely-rhino-f9r5l template="{\"name\":\"run-test\",\"inputs\":{},\"outputs\":{},\"metadata\":{},\"container\":{\"name\":\"runner\",\"image\":\"alpine\",\"command\":[\"sh\",\"-c\"],\"args\":[\"exit 0\"],\"resources\":{}}}" version="&Version{Version:v3.4.4,BuildDate:2022-11-29T16:49:53Z,GitCommit:3b2626ff900aff2424c086a51af5929fb0b2d7e5,GitTag:v3.4.4,GitTreeState:clean,GoVersion:go1.18.8,Compiler:gc,Platform:linux/amd64,}"
lovely-rhino-f9r5l: time="2023-01-24T07:39:44.478Z" level=info msg="Starting deadline monitor"
lovely-rhino-f9r5l: time="2023-01-24T07:39:46.479Z" level=info msg="Main container completed" error="<nil>"
lovely-rhino-f9r5l: time="2023-01-24T07:39:46.479Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
lovely-rhino-f9r5l: time="2023-01-24T07:39:46.479Z" level=info msg="No output parameters"
lovely-rhino-f9r5l: time="2023-01-24T07:39:46.479Z" level=info msg="No output artifacts"
lovely-rhino-f9r5l: time="2023-01-24T07:39:46.479Z" level=info msg="Deadline monitor stopped"
lovely-rhino-f9r5l: time="2023-01-24T07:39:46.479Z" level=info msg="stopping progress monitor (context done)" error="context canceled"
lovely-rhino-f9r5l: time="2023-01-24T07:39:46.479Z" level=info msg="Alloc=6098 TotalAlloc=12048 Sys=18386 NumGC=4 Goroutines=4"lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:54.519Z" level=info msg="Starting Workflow Executor" version=v3.4.4
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:54.522Z" level=info msg="Using executor retry strategy" Duration=1s Factor=1.6 Jitter=0.5 Steps=5
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:54.522Z" level=info msg="Executor initialized" deadline="2023-01-24 09:39:42 +0000 UTC" includeScriptOutput=false namespace=test podName=lovely-rhino-f9r5l-cowsay-100404247 template="{\"name\":\"cowsay\",\"inputs\":{\"parameters\":[{\"name\":\"ternary\",\"value\":\"FAILED\"},{\"name\":\"status\",\"value\":\"Succeeded\"}]},\"outputs\":{},\"metadata\":{},\"container\":{\"name\":\"\",\"image\":\"argoproj/argosay:v1\",\"command\":[\"sh\",\"-c\"],\"args\":[\"cowsay \\\"FAILED Succeeded\\\"\"],\"resources\":{}}}" version="&Version{Version:v3.4.4,BuildDate:2022-11-29T16:49:53Z,GitCommit:3b2626ff900aff2424c086a51af5929fb0b2d7e5,GitTag:v3.4.4,GitTreeState:clean,GoVersion:go1.18.8,Compiler:gc,Platform:linux/amd64,}"
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:54.522Z" level=info msg="Starting deadline monitor"
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:56.523Z" level=info msg="Main container completed" error="<nil>"
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:56.523Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:56.523Z" level=info msg="No output parameters"
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:56.523Z" level=info msg="No output artifacts"
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:56.523Z" level=info msg="Alloc=7745 TotalAlloc=11920 Sys=18386 NumGC=3 Goroutines=7"
```

### PR DESCRIPTION
**Pull request by jiachengxu**
Signed-off-by: Jiacheng Xu <xjcmaxwellcjx@gmail.com>

Fixes #10393
 
This PR fixes the issue that exit handler variables(https://argoproj.github.io/argo-workflows/variables/#exit-handler) `workflow.status` and `workflow.failures` don't get resolved correctly.

The root cause of this issue is that the above 2 parameters will only be added to the global parameter map when the exit handler completes. If the expressions that contains the above 2 parameters are evaluated when they don't exist, Expr will evaluate them according to the zero value of map (nil) (https://github.com/antonmedv/expr/issues/330), for example:

`{{= workflow.status == "Succeeded" ? "SUCCESSFUL": "FAILED" }}`

When `workflow.status` doesn't exist in the parameter map, it will be evaluated as `nil` and the above expression becomes:

`{{ nil == "Succeeded" ? "SUCCESSFUL" : "FAILED" }}`

It will return `FAILED`.

This PR skips the expression evaluation if the above parameters are not in the parameter map, and only evaluates the expressions when exit handlers complete, and parameters are added to the map.

Please do not open a pull request until you have checked ALL of these:

* [x] Create the PR as draft .
* [x] Run `make pre-commit -B` to fix codegen and lint problems.
* [x] Sign-off your commits (otherwise the DCO check will fail).
* [x] Use [a conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/) (otherwise the commit message check will fail).
* [x] "Fixes #" is in both the PR title (for release notes) and this description (to automatically link and close the issue).
* [x] Add unit or e2e tests. Say how you tested your changes. If you changed the UI, attach screenshots.
* [x] Github checks are green.
* [x] Once required tests have passed, mark your PR "Ready for review".

If changes were requested, and you've made them, dismiss the review to get it reviewed again.

### CODE REVIEW

File: test/e2e/testdata/workflow-hook-parameter.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+          - exit 0
+  hooks:
+    exit:
+      template: cowsay
+      arguments:
+        parameters:
+          - name: ternary
+            value: '{{= workflow.status == "Succeeded" ? "SUCCESSFUL" : "FAILED" }}'
+          - name: status
+            value: '{{= workflow.status }}'
```
Comment (jessesuen) [inline test/e2e/testdata/workflow-hook-parameter.yaml:33]: We can use a self-testing workflow 

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: workflow-hook-parameter
spec:
  entrypoint: run-test
  templates:
    - name: run-test
      container:
        name: runner
        image: 'argoproj/argosay:v2'
        command: ['sh','-c']
        args:
          - exit 0
    - name: cowsay
      inputs:
        parameters:
          - name: ternary
          - name: status
      container:
        image: 'argoproj/argosay:v2'
        command: ['bash','-c']
        args:
          - |
            echo "{{inputs.parameters.ternary}}"
            echo "{{inputs.parameters.status}}"
            [[ "{{inputs.parameters.ternary}}" = "{{inputs.parameters.status}}" ]]
  hooks:
    exit:
      template: cowsay
      arguments:
        parameters:
          - name: ternary
            value: '{{= workflow.status == "Succeeded" ? "Succeeded" : "Failed" }}'
          - name: status
            value: '{{= workflow.status }}'
```

Comment (jiachengxu) [reply test/e2e/testdata/workflow-hook-parameter.yaml]: Thank you! Done.

**Comment by sarabala1979** [comment]
@jiachengxu Do you think if workflow.status added into the map will resolve this issue?
https://github.com/argoproj/argo-workflows/blob/ab178bb0b36a5ce34b4c1302cf4855879a0e8cf5/workflow/controller/operator.go#L1755

**Comment by sarabala1979** [comment]
I think status is adding to globalparam. I think it not updating the latest when it execute the exit hanlder
https://github.com/argoproj/argo-workflows/blob/ab178bb0b36a5ce34b4c1302cf4855879a0e8cf5/workflow/controller/operator.go#L3630

**Comment by jiachengxu** [comment]
> Do you think if workflow.status added into the map will resolve this issue?

@sarabala1979 Thanks for your review!
I don't think that will resolve the issue, since as you mentioned, status is added to global params. and in the following example:
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: workflow-hook-parameter
spec:
  entrypoint: run-test
  templates:
    - name: run-test
      container:
        name: runner
        image: 'argoproj/argosay:v2'
        command: ['sh','-c']
        args:
          - exit 0
    - name: cowsay
      inputs:
        parameters:
          - name: ternary
          - name: status
      container:
        image: 'argoproj/argosay:v2'
        command: ['bash','-c']
        args:
          - |
            echo "{{inputs.parameters.ternary}}"
            echo "{{inputs.parameters.status}}"
            [[ "{{inputs.parameters.ternary}}" = "{{inputs.parameters.status}}" ]]
  hooks:
    exit:
      template: cowsay
      arguments:
        parameters:
          - name: ternary
            value: '{{= workflow.status == "Succeeded" ? "Succeeded" : "Failed" }}'
          - name: status
            value: '{{= workflow.status }}'
```
`'{{= workflow.status }}'` can get substituted but `'{{= workflow.status == "Succeeded" ? "Succeeded" : "Failed" }}'` cannot.
The reason is that, as you pointed out, `workflow.status` will only be added to the Globalparams map after `setGlobalRuntimeParameters`(https://github.com/argoproj/argo-workflows/blob/ab178bb0b36a5ce34b4c1302cf4855879a0e8cf5/workflow/controller/operator.go#L3630) is called, and if it is not called, and status is not added to the globalparams,
- `'{{= workflow.status }}'` will not be substituted since status is nil in the map.
- `'{{= workflow.status == "Succeeded" ? "Succeeded" : "Failed" }}'` will be substituted to `Failed` since status is nil, and the expression equals to `{{ nil == "Succeeded" ? "Succeeded" : "Failed" }}`, and will be substituted to `Failed`.

That's why we see different behaviors even if workflow.status is nil for the above cases.

What this PR is doing is that before we evaluate the expression, we check if the runtime parameters are in globalparams map, if they are not, we don't evaluate them(e.g., `'{{= workflow.status == "Succeeded" ? "Succeeded" : "Failed" }}'` will remain as it is), and once the `setGlobalRuntimeParameters` is called and the runtime parameters are added to globalparams map, we then evaluate the expression and substitute the values.

**Comment by kim-codefresh** [comment]
Hey @jessesuen @jiachengxu we cherry pick this fix into our fork that using version 3.4.4, but we still had issue - report failure when workflow status is success on exist handler on our workflows (not consist). 
Is this fix might be depended on version 3.4.5 somehow?

**Pull request merged**